homeassistant/components/systemmonitor/coordinator.py

```diff
@@ -0,0 +1,233 @@
+    "Package id 0",
+    "Physical id 0",
+    "radeon 1",
+    "soc-thermal 1",
+    "soc_thermal 1",
+    "Tctl",
+    "cpu0-thermal",
+    "cpu0_thermal",
+    "k10temp 1",
+]
+
+_dataT = TypeVar(
+    "_dataT",
+    bound=sdiskusage
+    | sswap
+    | svmem
+    | int
+    | float
+    | str
+    | tuple[float, float, float]
+    | datetime
+    | bool
+    | None,
+)
+
+
+class MonitorCoordinator(DataUpdateCoordinator[_dataT]):
+    """A System monitor Base Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, type_: str, argument: str) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN} {type_}_{argument}",
+            update_interval=DEFAULT_SCAN_INTERVAL,
+            always_update=False,
+        )
+        self._type = type_
+        self._argument = argument
+
+    async def _async_update_data(self) -> _dataT:
+        """Fetch data."""
+        return await self.hass.async_add_executor_job(self.update_data)
+
+    @abstractmethod
+    def update_data(self) -> _dataT:
+        """To be extended by data update coordinators."""
+
+
+class SystemMonitorDiskCoordinator(MonitorCoordinator[sdiskusage]):
+    """A System monitor Disk Data Update Coordinator."""
+
+    def update_data(self) -> sdiskusage:
+        """Fetch data."""
+        return psutil.disk_usage(self._argument)
+
+
+class SystemMonitorSwapCoordinator(MonitorCoordinator[sswap]):
+    """A System monitor Swap Data Update Coordinator."""
+
+    def update_data(self) -> sswap:
+        """Fetch data."""
+        return psutil.swap_memory()
+
+
+class SystemMonitorMemoryCoordinator(MonitorCoordinator[svmem]):
+    """A System monitor Memory Data Update Coordinator."""
+
+    def update_data(self) -> svmem:
+        """Fetch data."""
+        return psutil.virtual_memory()
+
+
+class SystemMonitorNetIOCoordinator(MonitorCoordinator[int]):
+    """A System monitor Network IO Data Update Coordinator."""
+
+    def update_data(self) -> int:
+        """Fetch data."""
+        counters = psutil.net_io_counters(pernic=True)
+        if self._argument in counters:
+            return counters[self._argument][IO_COUNTER[self._type]]
+        raise UpdateFailed(f"NIC {self._argument} could not be found")
+
+
+class SystemMonitorNetThroughputCoordinator(MonitorCoordinator[float | None]):
+    """A System monitor Network Throughput Data Update Coordinator."""
+
+    def __init__(self, hass: HomeAssistant, type_: str, argument: str) -> None:
+        """Initialize the coordinator."""
+        super().__init__(hass, type_, argument)
+        self._value: float | None = None
+        self._update_time: datetime | None = None
+
+    def update_data(self) -> float | None:
+        """Fetch data."""
+        counters = psutil.net_io_counters(pernic=True)
+        if self._argument in counters:
+            counter = counters[self._argument][IO_COUNTER[self._type]]
```

Comment:
I don't think argument and type belongs on the coordinator, besides the disk coordinator which needs the argument. For the rest the argument and type are specific for each entity not for each coordinator. The coordinators should just return the measurement from the measurement api call. Then it's up to each entity and its entity description to parse the measurement result and set a state.

The `psutil.net_io_counters(pernic=True)` call just needs to go in one coordinator and all the net io sensors can use that coordinator.